src/witan/send/adroddiad/ncy.clj

```diff
@@ -1,7 +1,298 @@
+ (-> (tc/concat (tc/dataset {:id  1 :calendar-year [2000 2001 2002] :academic-year      nil     })
+                (tc/dataset {:id  2 :calendar-year [2000 2001 2002] :academic-year [nil   1 nil]})
+                (tc/dataset {:id  3 :calendar-year [2000 2001 2002] :academic-year [nil nil   2]})
+                (tc/dataset {:id  4 :calendar-year [2000 2001 2002] :academic-year [nil nil  -4]})
+                (tc/dataset {:id  5 :calendar-year [2000 2001 2002] :academic-year [ 20 nil nil]})
+                (tc/dataset {:id  6 :calendar-year [2000 2001 2002] :academic-year [  0   1   2]}))
+     (impute-nil-ncy)
+     (tc/drop-columns :academic-year-imputation)
+     (tc/order-by [:id :calendar-year])) :=
+ (-> (tc/concat (tc/dataset {:id  1 :calendar-year [2000 2001 2002] :academic-year      nil     })
+                (tc/dataset {:id  2 :calendar-year [2000 2001 2002] :academic-year [  0   1   2]})
+                (tc/dataset {:id  3 :calendar-year [2000 2001 2002] :academic-year [  0   1   2]})
+                (tc/dataset {:id  4 :calendar-year [2000 2001 2002] :academic-year [ -6  -5  -4]})
+                (tc/dataset {:id  5 :calendar-year [2000 2001 2002] :academic-year [ 20  21  22]})
+                (tc/dataset {:id  6 :calendar-year [2000 2001 2002] :academic-year [  0   1   2]})))
+
+ 
+ "Where imputation occurs the source should be described in the `:academic-year-imputation` column"
+ (-> (tc/dataset {:id  7 :calendar-year [2000 2001] :academic-year [nil   1]})
+     (impute-nil-ncy)) :=
+ (-> (tc/dataset {:id  7 :calendar-year [2000 2001] :academic-year [  0   1] :academic-year-imputation ["Imputed from NCY=1 in 2001" nil]}))
+ 
+ 
+ "Where column names are specified they should be used, including specification of a separate column for the NCY after imputation."
+ (-> (tc/dataset {:id  8 :cy [2000 2001] :ncy [nil   1]})
+     (impute-nil-ncy {:cy :cy :ncy :ncy :ncy-imputed :ncyi :ncy-imputation :ncyi-desc})) :=
+ (-> (tc/dataset {:id  8 :cy [2000 2001] :ncy [nil   1]  :ncyi [  0   1] :ncyi-desc ["Imputed from NCY=1 in 2001" nil]}))
+
+
+ "Should also be able to specify options as keyword parameters (rather than a map), and the algorithm should work with non-keyword column names"
+ (-> (tc/dataset {'id  9 "SEN2 census year" [2000 2001] 'NCY [nil   1]})
+     (impute-nil-ncy :id 'id :cy "SEN2 census year" :ncy 'NCY :ncy-imputed 'NCY :ncy-imputation 'ncy-imputation-desc)) :=
+ (-> (tc/dataset {'id  9 "SEN2 census year" [2000 2001] 'NCY [  0   1] 'ncy-imputation-desc ["Imputed from NCY=1 in 2001" nil]}))
+
+ 
+ "With multiple possible ref. records, the one with lowest census/calendar-year should be used (:id 9 & 10).
+  If there are multiple such records, then the one with smallest NCY is chosen (:id 11).
+  Note that if pupils do not progress at a rate of one NCY per census/calendar-year then this can result in
+  discontinuities if a CYP repeats or goes back a NCY and has missing NCY in subsequent years (:id 12 & 13)."
```

Comment:
I like this test as it explains the expected if possibly surprising behaviour in some circumstances